dpcpp/matrix/bccoo_kernels.dp.cpp

```diff
@@ -0,0 +1,1241 @@
+}
+
+GKO_ENABLE_DEFAULT_HOST(abstract_absolute_inplace, abstract_absolute_inplace);
+
+
+/**
+ * The device function of BCCOO compute_absolute
+ *
+ * @param nnz  the number of nonzeros in the matrix
+ * @param num_blks  the number of blocks in the matrix
+ * @param block_size  the number of nonzeros in each block
+ * @param num_lines  the maximum round of each warp
+ * @param chunk_data_src  the array where the data of source are
+ * @param offsets_data_src  the array where the offset of each block of source
+ * is
+ * @param types_data_src  the array where the sorce type of each block of source
+ * is
+ * @param cols_data_src  the array where the initial column of each block of
+ * source is
+ * @param rows_data_src  the array where the initial row of each block of source
+ * is
+ * @param chunk_data_res  the array where the data of result are
+ * @param offsets_data_res  the array where the offset of each block of result
+ * is
+ * @param types_data_res  the array where the sorce type of each block of result
+ * is
+ * @param cols_data_res  the array where the initial column of each block of
+ * result is
+ * @param rows_data_res  the array where the initial row of each block of result
+ * is
+ * @param comp_abs  the function to apply to obtain the solution
+ *
+ * @tparam subgroup_size   size of the subgroup
+ * @tparam ValueType  type of values stored in the matrix
+ * @tparam IndexType  type of matrix indexes stored in the structure
+ * @tparam Closure  type of the function used to write the result
+ */
+template <int subgroup_size = config::warp_size, typename ValueType,
+          typename IndexType, typename Closure>
+void absolute_kernel(const IndexType nnz, const IndexType num_blks,
+                     const IndexType block_size, const IndexType num_lines,
+                     const uint8* __restrict__ chunk_data_src,
+                     const size_type* __restrict__ offsets_data_src,
+                     const uint8* __restrict__ types_data_src,
+                     const IndexType* __restrict__ cols_data_src,
+                     const IndexType* __restrict__ rows_data_src,
+                     uint8* __restrict__ chunk_data_res,
+                     size_type* __restrict__ offsets_data_res,
+                     uint8* __restrict__ types_data_res,
+                     IndexType* __restrict__ cols_data_res,
+                     IndexType* __restrict__ rows_data_res, Closure comp_abs,
+                     sycl::nd_item<3> item_ct1)
+{
+    const IndexType column_id = item_ct1.get_group(1);
+    const IndexType start_blk = item_ct1.get_group(2);
+    const IndexType jump_blk = item_ct1.get_group_range(2);
+
+    const IndexType start_in_blk =
+        item_ct1.get_local_id(1) * subgroup_size + item_ct1.get_local_id(2);
+    const IndexType jump_in_blk =
+        item_ct1.get_local_range().get(1) * subgroup_size;
+
+    if (start_blk == 0) {
+        offsets_data_res[0] = 0;
+    }
```

Comment:
same comments in cuda_hip. using start_in_blk to avoid the multiple write.

Reply:
Done!!